lib/site_cache.rb

```diff
@@ -0,0 +1,213 @@
+# frozen_string_literal: true
+
+require "monitor"
+
+# This is an LRU cache in two dimensions. There are limits on:
+#
+# * The global size of the cache,
+# * The per-site size of the cache,
+#
+# In order to enforce these limits, we need be able to efficiently get the next
+# least recently used item globally and per site. To do this, there are two
+# structures:
+#
+# `@flattened' is a Hash from [site, key] -> value
+# `@nested' is a Hash from site -> Hash from key -> value
+#
+# In both cases, we exploit the ordered property of ruby hashes. `#shift` on a
+# hash pops the least recently inserted key-value pair. If all accesses delete
+# the item and reinsert it, this gives us LRU order.
+class SiteCache
+  include MonitorMixin
+
+  def initialize(max_global_size, max_size_per_site)
+    mon_initialize
+
+    raise ArgumentError.new(:max_global_size) if max_global_size < 1
+    raise ArgumentError.new(:max_size_per_site) if max_size_per_site < 1
```

Comment:
Do we also have to ensure that `max_global_size` is greater than `max_size_per_site`?

Reply:
It would still work if it wasn't, but sure.